### PR TITLE
WIP: SonarQube configuration

### PR DESCRIPTION
**Pull request by alixwar**
## Overview

Added SonarQube (static code analysis) configuration. Tested on a local instance of SonarQube 6.5 (build 27846). Clover configuration, e.g. code coverage, has not been verified (requires license). Run with default parameters (sonar.host.url=localhost:9000, sonar.login=admin, sonar.password=admin) using command: gradlew sonarqube

Next step:
- Register free account at https://about.sonarcloud.io/ and follow simple tutorial steps to integrate with CI

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by alixwar** [comment]
@sormuras @sbrannen This shouldn't be in the release notes, right?

**Comment by sbrannen** [comment]
> > * Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
> 
> @sormuras @sbrannen This shouldn't be in the release notes, right?

Correct

**Comment by alixwar** [comment]
@sbrannen I have a dilemma, the build script in this branch is missing server configuration and login credentials to https://sonarcloud.io so the build fails. Someone from the team needs to acquire credentials. If you could do that for me I can put the missing parameters into this branch so that the build passes

**Comment by marcphilipp** [comment]
@alixwar Could you please set it up for your fork first?

**Comment by alixwar** [comment]
> @alixwar Could you please set it up for your fork first?

Sure thing. But I still need credentials that only the JUnit team can get for me on: https://sonarcloud.io/

**Comment by marcphilipp** [comment]
How so? In my understanding, you can create a project for your fork without needing our credentials. Isn't that so?

**Comment by alixwar** [comment]
> How so? In my understanding, you can create a project for your fork without needing our credentials. Isn't that so?

I can make it work locally with a local installation of SonarQube. But if I want it to run in the cloud I need an account at sonarcloud.io. This is free to use for open source projects but I don't think that my fork would quality as an open source project in itself.

**Comment by marcphilipp** [comment]
I could supply you with credentials and encode them as Travis variables. However, they wouldn't work on builds for branches not from the core repo. I don't see a reason why sonarcloud.io would treat your fork differently from this repo. Could you please double-check?

**Comment by alixwar** [comment]
@marcphilipp Here is the result: https://sonarcloud.io/dashboard?id=alixwar_junit5

I can fix the missing code coverage metric if I can find out where the exec files end up (I cannot build the project locally)

**Comment by marcphilipp** [comment]
Looking good! We merge the exec files and there will be one in `build/jacoco/jacocoMerge.exec` in the root project.

**Comment by alixwar** [comment]
@marcphilipp The coverage configuration didn't work: https://github.com/junit-team/junit5/pull/1008/commits/894e4bb612dc6d904ae563b9f447fe630ccbd1a2 

I could try different things such as adding "" or "." but I don't want to clutter the GIT log if there is a simpler explanation (such as .exec files need to be explicitly exported from the previous job somehow). Do you see any obvious mistake (I'm not familiar with travis-ci)?

**Comment by aepfli** [comment]
@alixwar maybe it helps to provide `-i` to the gradlew task -> this will increase the output, and will mostlikely ease up debugging for you. the sonarqube plugin is quiet a chatty one :D

**Comment by aepfli** [comment]
Additionally the sonarqube plugin will automatically create 'sonar' modules out of subprojects, as it is used by junit5. hence that i think that the usage of an overall exec file might be tricky, because the module will always act as its own root. But i highly recommend to use the -i flag, because this will enlighten you, and will ease up this investigation.

for more details: https://docs.sonarqube.org/display/SCAN/Analyzing+with+SonarQube+Scanner+for+Gradle

**Comment by marcphilipp** [comment]
@alixwar Don't worry about cluttering the Git log. We usually squash all comits before merging them to `master` anyway.

**Comment by marcphilipp** [comment]
The way Travis works, the exec file won't be present in the following build. Please try merging the "Coverage" and "Code analysis" jobs into a single job with multiple scripts.

**Comment by aepfli** [comment]
> The way Travis works, the exec file won't be present in the following build. Please try merging the "Coverage" and "Code analysis" jobs into a single job with multiple scripts.

i have to say  if you merge those two, i would not recommend to put it into multiple scripts. If you do it like that the sonarqube gradle plugin will automatically detect the jacoco report based on the test execution, and you can skip the part with configuration. See https://docs.sonarqube.org/display/SCAN/Analyzing+with+SonarQube+Scanner+for+Gradle#AnalyzingwithSonarQubeScannerforGradle-AdditionaldefaultswhenJaCoCopluginisapplied

**Comment by alixwar** [comment]
SonarQube handles short-lived and long-lived branches differently (the major analysis is performed on long-lived branches only). My thinking is to get the basic analysis working first (e.g. long-lived branch analysis) and later improve by adding short-lived branch analysis as well, if requested.

As for the script part, I've made a suggested approach for now to avoid redundant jobs (some code duplication as you can see): https://github.com/junit-team/junit5/pull/1008/commits/bc5c5c4d56404f0ed1e6fa97b0d45e6b1af1e27e

**Comment by aepfli** [comment]
> SonarQube handles short-lived and long-lived branches differently (the major analysis is performed on long-lived branches only). My thinking is to get the basic analysis working first (e.g. long-lived branch analysis) and later improve by adding short-lived branch analysis as well, if requested.

That would have been my next question - and i like your approach regarding that.

What i meant with 

> i have to say if you merge those two, i would not recommend to put it into multiple scripts. If you do it like that the sonarqube gradle plugin will automatically detect the jacoco report based on the test execution, and you can skip the part with configuration.

is that if you but the whole execution into one command like:

`  - ./gradlew --scan --stacktrace --warning-mode=all -PenableJaCoCo build jacocoRootReport sonarqube --info -Dsonar.host.url=https://sonarcloud.io -Dsonar.login=6405ce221514f081109da1c961d11ab5c5284038 -Dsonar.projectKey=alixwar_junit5 -Dsonar.organization=alixwar-github`

You do not need to care about the jacoco report anymore. The reason is, that the jacoco plugin is applied for test execution and is configured. And if this is the case, the sonarqube plugin, will take the reference to the JaCoCo report out of the build script automatically, and you will not need to configure it by yourself.

I do not know if this would be suitable for the junit project, but it will reduce the effort for configuration and will make it easier to apply this to future projects.

**Comment by alixwar** [comment]
@aepfli Your suggestion did the trick. There is now coverage (at least some)!

**Comment by aepfli** [comment]
@alixwar i saw that also, and there are some parts with are worth investigation.

First of all, i am not sure what the mergeJacocoReport is doing, it might take the existing reports and merge them so that they are not present anymore, hence that you might lack coverage information - just a wild guess. It would be fixed by running the JacocoReport target after sonarqube. But that is really a wild guess, because in point 2 you will see a logoutput where he finds data, but can not match it to the data. But be aware, that just changing the ordering in the script might not generate the expected output. Gradle might reorder those tasks.

Second i saw some logs, that the information in the *.exec file is not correlating to the class files. This is strange, but would also need some kind of attention, because it can also effect the coverage report.
from https://api.travis-ci.org/v3/job/498925125/log.txt:
```
Sensor JaCoCoSensor [java]
Analysing /home/travis/build/junit-team/junit5/junit-jupiter-params/build/jacoco/test.exec
The following class(es) did not match with execution data:
> 'org/junit/jupiter/params/provider/EnumArgumentsProvider'
> 'org/junit/jupiter/params/aggregator/ArgumentAccessException'
> 'org/junit/jupiter/params/provider/CsvParsingException'
> 'org/junit/jupiter/params/ParameterizedTestMethodContext$ResolverType$1'
> 'org/junit/jupiter/params/ParameterizedTestMethodContext$ResolverType$2'
> 'org/junit/jupiter/params/converter/FallbackStringToObjectConverter$IsFactoryConstructor'
> 'org/junit/jupiter/params/converter/FallbackStringToObjectConverter$IsFact
```

**Comment by alixwar** [comment]
> First of all, i am not sure what the mergeJacocoReport is doing, it might take the existing reports and merge them so that they are not present anymore, hence that you might lack coverage information - just a wild guess. It would be fixed by running the JacocoReport target after sonarqube

But if this was the case, then I would need an explicit jacoco report configuration again as mentioned here: https://github.com/junit-team/junit5/pull/1008#issuecomment-467572138

**Comment by aepfli** [comment]
no i dont think so, right now the execution is:

1. tests
2. JacocoMerge
3. sonarqube

but if you change the order to
1. tests
2. sonarqube
3. JaCoCoReport

you will remove that issue - but as said this is a wild guess. I can try to debug locally and give you more information. The thing with the Gradle ordering can be achieved by some gradle directives like <taskX>.mustRunAfter "<taskY>" in the after evaluate block. Just to clarify those are just guesses, i am putting here, but i will most likely try to debug locally tonight or tomorrow

**Comment by aepfli** [comment]
@alixwar is it okay for you, if i help on this task, and if so, is it okay for you if i use your token locally?

**Comment by alixwar** [comment]
@aepfli Of course, go right ahead!

**Comment by aepfli** [comment]
so short summary of things i found out:

1. it is not jacoco merge - this is really just copying the files

2. the coverage so far is working, for all the execs. maybe i am missing some target, but when i check my local exec files after execution, they sometimes also do not contain any information

3. Junit jupiter params has some kind of issue with matching the files from the jacoco report to the actual files, hence that, sonarqube can not compute any coverage for that plugin. This might relate to different java versions used. But i had not time do deeper look into the gradle structure, and what the shadowjar is suppose to be doing. (but that is something on the side)

So right now, i hit a wall. and my time this week is limited, i will try to find some time. But for now my goal is getting a better understanding of the gradle structure, and shadow jar :D

**Comment by alixwar** [comment]
@marcphilipp Do you consider the part of gathering code coverage a must have to merge this PR? Otherwise we could disable the Jacoco configuration for now (so we don't get skewed coverage numbers) and try to sort this out in a new PR.

**Comment by marcphilipp** [comment]
@alixwar I think it's fine to add coverage info in a subsequent PR. Could you please remove it for now as you suggested?

**Comment by alixwar** [comment]
> @alixwar I think it's fine to add coverage info in a subsequent PR. Could you please remove it for now as you suggested?

I have now pruned the jacoco part of the CI step

**Comment by aepfli** [comment]
> I have now pruned the jacoco part of the CI step

but does it make sense to have the codecov state in there? i mean now you do not have any coverage generated. This might be a flaw in the process right now.

i think you should leave the jacoco part in there, else you will break the codecov, as it might not find the coverage info. - hence that maybe overriding in the settings the `sonar.jacoco.reportPaths` to point to wrong jacoco reports, or exclude everything from coverage detection with `sonar.coverage.exclusions=**/*` might be more useful.

**Comment by marcphilipp** [comment]
Agreed, let's keep Codecov for now.

File: .travis.yml [hunk 1]
```diff
@@ -47,9 +47,15 @@ stages:
 jobs:
   include:
     - name: "Coverage"
+      if: branch != master
```
Comment (marcphilipp) [inline .travis.yml]: Let's do this always and remove JaCoCo from the job below.

Comment (alixwar) [reply .travis.yml]: OK

**Comment by marcphilipp** [comment]
Thanks, @alixwar! I'll take it from here and replace the credentials etc. with the ones for this repo.

**Comment by aepfli** [comment]
May I suggest to move some settings to the Gradle script - like coverage exclusion etc.

Furthermore, what we tend to do is, placing them into an own sonar-project.properties file and load this. This allows to even use different Gradle scanner as most of them are looking for such a file/configuration on startup.

**Comment by marcphilipp** [comment]
A separate properties file sounds like a good idea but we should put it under `src` or `src/sonar` so we don't (further) "pollute" the root directory.

**Comment by alixwar** [comment]
@marcphilipp Would you like me to take care of the transfer of properties to a properties file or do you want to do it?

If using a non-standard path (i.e. not in the root directory) for sonar-project.properties you need to explicitly reference the file using:
`-Dproject.settings`

Source: https://docs.sonarqube.org/display/SCAN/Advanced+SonarQube+Scanner+Usages

**Comment by marcphilipp** [comment]
I'll have a look.

**Comment by aepfli** [comment]
Sorry that i alway scoop in here, but i also took a look at the azure pipeline scripts, and i thought it would be a great idea to create an additional gradle target called `coverage` which will be setting jacocoEnabled and run build and JacocoRootReport

on the long run, it will be easier to handle and maintain if you have different build scripts, as you just adapt the gradle task, and it will be automatically used within all the build scripts.

If you (@marcphilipp) think that this is out of scope of this task, i can create another issue tackeling this topic.

**Comment by marcphilipp** [comment]
First of all, I appreciate the time that you've already put into this, @alixwar and @aepfli! However, I think that we should take a step back and reconsider whether SonarQube will actually be useful to the team. I looked through the initial analysis on https://sonarcloud.io/dashboard?id=alixwar_junit5 and have yet to find anything I'd consider truly useful. Please don't take this personally, this is purely about the tool, not about your work. I'm sure the profile could be tweaked to remove useless code smells. Even so, rulesets change and we'd have to do this repeatedly. I feel that this would be a considerable effort that we should rather spend working on JUnit. We discussed this in our last team meeting and all agree. Thus, unless you can convince me otherwise, I'm inclined to close this PR without merging it.

**Comment by aepfli** [comment]
I am fine, I just put in here some suggestions, and I think I already figured out why the coverage was wrong ;)

Well in this project there are real good gatekeepers and the overall quality of pull request is great. So I can understand your decision and I do not see the need. Although for less experienced developers such a tool can have a high impact and be a great source for improvements. But as I said i also do not see a lot of benefit due to the high quality and the good gatekeeping.

**Comment by alixwar** [comment]
I view the tool as a safety net, which obviously has failed to prove its use in this particular project with its current team (kudos!) and processes.

Judging the tool as useless by its current (lack of) findings doesn't prove much though:
If there are no high severity issues in the code base right now it makes perfect sense that the tool does not find any high severity rule violations. But I can see your point that if the current findings really are false positives (19 bugs, 1 vulnerability) then it could be considered a waste of time to have a process where these findings need to be dealt with.

I also understand why you don't want to clutter the project with configurations for tools that you don't actively use so I reluctantly accept your decision without grudge :)

**Comment by marcphilipp** [comment]
@alixwar I'm glad! 🙂

**Pull request closed without merging**